examples/gguf-split/gguf-split.cpp

```diff
@@ -0,0 +1,510 @@
+#include "llama.h"
+#include "ggml.h"
+#include "common.h"
+
+#include <algorithm>
+#include <cmath>
+#include <cstdint>
+#include <cstdlib>
+#include <fstream>
+#include <ios>
+#include <string>
+#include <vector>
+
+#include <stdio.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <string.h>
+#include <unistd.h>
+
+enum split_operation : uint8_t {
+  SPLIT_OP_SPLIT,
+  SPLIT_OP_MERGE,
+  SPLIT_OP_UPLOAD,
+};
+
+static const char * const LLM_KV_GENERAL_SPLIT_I_SPLIT = "general.split";
+static const char * const LLM_KV_GENERAL_SPLIT_N_SPLIT = "general.split_count";
+
+static const int SPLIT_FILENAME_MAX = 256;
+
+static const char * const SPLIT_FILENAME_FORMAT = "%s-%05d-of-%05d.gguf";
+
+struct split_params {
+  split_operation operation = SPLIT_OP_SPLIT;
+  int n_split_tensors = 128;
+  std::string input;
+  std::string output;
+};
+
+static void split_print_usage(const char * executable) {
+    const split_params default_params;
+    printf("\n");
+    printf("usage: %s [options] GGUF_IN GGUF_OUT\n", executable);
+    printf("\n");
+    printf("Apply a GGUF operation on IN to OUT.");
+    printf("\n");
+    printf("options:\n");
+    printf("  -h, --help            show this help message and exit\n");
+    printf("  --version             show version and build info\n");
+    printf("  --split               split GGUF to multiple GGUF (default)\n");
+    printf("  --split-tensors-size  size of tensors in each split: default(%d)\n", default_params.n_split_tensors);
```

Comment:
I don't think `--split-tensors-size` is practical because:
- This param is more about "max **number** of tensor", not their size (in bytes)
- Most user doesn't know how many tensors are there in a file
- They don't know either each tensor weights how many KB or MB

I prefer having `--split-max-size` and we accepts something like `128M` or `5G`. The `should_split()` can see if `current_split_file_offset + next_tensor_size > split_max_size` , that will make sure each splits never passes a threshold. The tricky part I think is how to calculate number of splits from the beginning.

`--split-max-size` should also match the behavior of huggingface, since it have a limit of 36G per file

Reply:
Good point. I just tested it with a llama2 70B, with `--split-tensors-size 32`, it generates 23 files varying from `3.7G` to `6.6G` and I think it's acceptable for a first version. I introduced in the code the `split_strategy` struct to later introduce other strategy like softlimit split file size you propose. I believe it can be done in another PR.

Reply:
I suggest that we can kinda "hide" `--split-tensors-size` in the first version (still leave it in the code, but do not mention in help) - This is just to prevent someone rely on this and we make breaking changes for them in the future.

Reply:
Sorry I assumed that we will no longer have `--split-tensors-size` in the future. It's still nice to keep I think, but rename to `--split-max-tensors` would be easier to understand.